libs/payments/cart/src/lib/cart.service.ts

```diff
@@ -92,16 +96,98 @@ export class CartService {
+          );
+          const invoice = subscription.latest_invoice
+            ? await this.invoiceManager.retrieve(subscription.latest_invoice)
+            : undefined;
+          if (invoice) {
+            switch (invoice.status) {
+              case 'draft':
+                await this.invoiceManager.delete(invoice.id);
+                break;
+              case 'open':
+              case 'uncollectible':
+                await this.invoiceManager.void(invoice.id);
+                break;
+              case 'paid':
+                throw new CartError('Paid invoice found on failed cart', {
+                  cartId,
+                  stripeCustomerId: cart.stripeCustomerId,
+                  invoiceId: invoice.id,
+                });
+            }
+          }
+
+          const paymentIntent =
+            await this.subscriptionManager.getLatestPaymentIntent(subscription);
+          if (paymentIntent) {
+            switch (paymentIntent.status) {
+              case 'requires_payment_method':
+              case 'requires_capture':
+              case 'requires_confirmation':
+              case 'requires_action':
+              case 'processing':
+                await this.paymentIntentManager.cancel(paymentIntent.id);
+                break;
+              case 'succeeded':
+                throw new CartError(
+                  'Succeeded payment intent found on failed cart',
+                  {
+                    cartId,
+                    stripeCustomerId: cart.stripeCustomerId,
+                    paymentIntentId: paymentIntent.id,
+                  }
+                );
+            }
+          }
+
+          await this.subscriptionManager.cancel(cart.stripeSubscriptionId);
+        }
+
+        // Remove Stripe Customer if they have no other subscriptions
+        let deleteStripeCustomer = false;
+        if (cart.stripeCustomerId) {
+          const activeSubscriptions = (
+            await this.subscriptionManager.listForCustomer(
+              cart.stripeCustomerId
+            )
+          ).filter(
+            (subscription) => subscription.id !== cart.stripeSubscriptionId
+          );
+          if (activeSubscriptions.length === 0) {
+            deleteStripeCustomer = true;
+            await this.customerManager.delete(cart.stripeCustomerId);
+          }
+        }
+
+        // Remove PayPal Customer if they have no other subscriptions
+        if (deleteStripeCustomer && cart.uid) {
+          await this.paypalCustomerManager.deletePaypalCustomersByUid(cart.uid);
+        }
+
+        // Remove AccountCustomer when Stripe Customer is deleted
+        const accountCustomer = cart.uid
+          ? await this.accountCustomerManager.getAccountCustomerByUid(cart.uid)
+          : undefined;
+        if (accountCustomer && deleteStripeCustomer) {
+          await this.accountCustomerManager.deleteAccountCustomer(
+            accountCustomer
+          );
+        }
```

Comment:
Why delete the account customer/customer -- Avoiding deleting the customer/accountCustomer was one of the things we wanted to change for SP3 AFAIK.

Reply:
The ticket explicitly details cleaning up these objects - https://mozilla-hub.atlassian.net/browse/FXA-10628